### DISCUSSION TITLE
X Twist Setup Guide

### DISCUSSION THREAD

**Discussion by banhathome** [bassamanator/Sovol-SV06-firmware#81] [Show and tell]
In Dec 2021, Marlin added a feature called "X-axis Twist Compensation": MarlinFirmware/Marlin#23238

At time of writing, Sovol STILL hasn't included this feature in their SV06 firmware. Fortunately, some folks like [hillsoftware have produced an updated firmware](https://github.com/hillsoftware/sv06) for the SV06 to include this feature.

At the moment, Klipper does not have this feature. There's been a PR open for it since Feb, but it still hasn't been merged: Klipper3d/klipper#6149

X Twist is probably one of the biggest issues that SV06 owners face, and I think it's a mandatory feature for owning this printer. Because of this, I have provided a guide on how you can get ahead of the Klipper and start using it today.

## Notes

- These instructions may quickly become obsolete when the PR or Klipper get updates.
- This is a software approach to fixing X Twist. But you should opt for a physical/mechanical fix first, ESPECIALLY if you're getting wild numbers for the compensation.
- I use OctoPi/OctoPrint so these instructions are written primarily for that. Feel free to let me know in the replies if you're having trouble adapting it to your environment.
  - I also have another post to provide some tips on setting up this repo's Klipper config for OctoPi: https://github.com/bassamanator/Sovol-SV06-firmware/discussions/78 
- I use an SV06. The instructions should be similar for the Plus but you'll need to update the values in your `printer.cfg` section.

## Instructions

- Find where your Klipper is installed. For me, it's in `~/klipper`.
  - You probably want to make sure your Klipper is up to date as well. But it most likely won't be necessary to update.
- Assuming you pulled your Klipper code from Git, run this in your terminal:
  
  ```bash 
  cd ~/klipper
  git fetch origin pull/6149/head:x-twist-compensation-simple
  git checkout x-twist-compensation-simple
  ```
- Run `sudo service klipper restart` to pick up these new changes.
- Update your `printer.cfg` to include the `[axis_twist_compensation]` setting. Here is my config:

  ```cfg
  [axis_twist_compensation]
  calibrate_start_x: 27
  calibrate_end_x: 195
  calibrate_y: 115
  ```

- Run `RESTART` in your OctoPrint terminal to pick up the new config changes.

- Run `G28` in your OctoPrint terminal to home your printer
- Run `AXIS_TWIST_COMPENSATION_CALIBRATE` in your OctoPrint terminal to begin the process. Follow the instructions.
- Run `SAVE_CONFIG` to save your axis twist compensation values.
- Run `DO_PROBE_CALIBRATE` to adjust your Z-Offset again. `SAVE_CONFIG` when you're done.
- (If you're not using KAMP) `DO_CREATE_MESH` to bed mesh. `SAVE_CONFIG` when it's done.
- You're done. Run a 1st layer bed level print and you'll find it printing a lot more evenly now.

**Comment by banhathome** [comment]
@bassamanator i remember seeing a post where you said you had to update your bed mesh manually to fix your x twist. Hopefully this helps

**Comment by bassamanator** [comment]
I spoke to the author of this PR back in March while I was hunting for a solution for my x-twist problem (which I've resolved via manually editing the mesh, very doable but time-consuming, and I don't consider it a proper solution). After a few months, I didn't think the PR would be merged, but it looks like it's going to actually happen!

I never got around to trying this, because I didn't have step-by-step instructions, like the kind you've posted above @chrisngobanh 😁. Plus I'm always working on one thing or another on this printer, I can't do it all. Will try this soon!

Good to see some enthusiasm on this repo! Pinned!

**Comment by bassamanator** [comment]
I can't believe how well this worked! BTW this works regardless of the interface, Mailsail, Octoprint, etc.

Minor changes on my end, I think it might provide a more 'balanced' view of the build plate, but I could be wrong:
```
[axis_twist_compensation]
calibrate_start_x: 27
calibrate_end_x: 195
calibrate_y: 115 # this keeps the probing point in line with the center screw
```

Hopefully this feature will be merged. I'll link to this discussion in the README.

Thanks for finding this gem @chrisngobanh!

**Reply by banhathome** [reply]
Updated the OP with your numbers. The ones I had originally were just something I threw together, so I trust your settings a lot more than mine.

**Comment by Catfisher4** [comment]
How do we know if we have an X-twist rather than an unlevel bed? Is there a surefire way to determine that you have X-twist? Sorry if this has been answered many times, but I didn't find answers on the original X-twist discussion

**Reply by bassamanator** [reply]
Well, I did the silicone mod, and manually levelled the bed. But my bed mesh continued to show that the left side was `lower`. This is a clue that something else is wrong here. If you've verified physically that the bed is level, and the mesh that you generate continues to tell you the opposite, this is one indicator that something is very off.

**Reply by banhathome** [reply]
Most guides will provide you with tips on how to level the bed in order to fix your prints. But if you find yourself still not getting level 1st layers, then it's possible that your X-axis is twisted.  X-Axis Twist is a very common problem in a lot of printers nowadays, especially with a budget printer like the SV06. 

Normally your probe is supposed to be able to help you compensate for any uneven or warped beds. But with a twisted X-axis, you may as well remove that probe and throw it in the trash because the data from it will be unusable until you account for the twist. Here's a Youtube video to demonstrate why: https://www.youtube.com/watch?v=rPU4ZAPJTNc

It took me a long time to realize that my X-axis was twisted; I thought I just wasn't leveling the bed correctly or I just didn't configure the bed mesh generation correctly. For me, I ran 1st layer bed leveling prints and consistently noticed that my left side would print super close while the right side would look like there was too much distance between the nozzle and bed. This is something that should be fixed via generating the bed mesh, which is why I was scratching my head figuring out why it wasn't working.

In the end, it wouldn't hurt just to give it a shot. Even if you're getting level prints, your printer probably still has a slight twist that could be accounted for. When it gets merged into Klipper, I think running the X-Axis Twist Compensation should be included in the setup instructions in the the README.

**Comment by bassamanator** [comment]
Merged into Klipper! 🎇

Will adjust configuration to include `axis_twist_compensation` section. 

It would help if any `SV06 Plus` users out there could create a PR with the coordinates of the bed screws, or just paste them here. 
```
[bed_screws]
screw1_name: front left
screw1: x, y
screw2_name: front right
screw2: x, y
screw3_name: back right
screw3: x, y
screw4_name: back left
screw4: x, y
screw5_name: center
screw5: x, y
```

**Reply by bassamanator** [reply]
I actually need to merge KAMP into 3 other branches. 

@chrisngobanh if you're up for it, feel free to create a PR for `axis_twist_compensation` whenever you have time, though don't stress about it.

**Reply by banhathome** [reply]
Awesome! Looks like your endorsement may have been the reason it finally got merged 😂 

I can definitely add it in when I get some free time. As I mentioned before, I think running the axis_twist_compensation should be included in the README's setup steps for SV06 Klipper.

**Reply by bassamanator** [reply]
lol I wish I was that well known in the larger Klipper community.

Take your time, and if you want to add a short write-up (basically the steps you have up there) to the `FAQ` section about the steps, feel free to do so. Maybe the prompt should be: `I have X-twist. How do I setup axis_twist_compensation?`. Unfortunately, this issue is quite common with this printer. Again, no rush. Thanks.

**Comment by asusdk** [comment]
> Will adjust configuration to include axis_twist_compensation section.

Does this mean it is already included, or do i have to change something in the config ?

Thanks in Advance

**Reply by bassamanator** [reply]
~~These change have not been included in the repo as of yet.~~
~~The master branch holds these changes.~~
Included on all branches. Unsure if settings chosen for `sv06-plus` are optimal.

What printer do you have?

**Comment by Trilis29** [comment]
Tried calibrating, but my left side was WAY too high after doing x twist. I am thinking that my bed may be  not level to table. Maybe firstly I should level bed with spirit level, then gantry, then probe, then try mesh and last resort x-axis?

**Reply by bassamanator** [reply]
You need not level the bed using a level. 

`axis_twist_compensation` is needed **only** for those that have X twist. 

Run through the initial steps on the everything sv06 repo.

**Reply by Trilis29** [reply]
I don't know if I have twist, wanted just to check out :) getting 0.8 range.

**Comment by adrianjdebono** [comment]
I dunno if its any help but I measured SV06 Plus bed screw positions. I did this with the probe directly above the hole is this correct?

[bed_screws]

screw1: Outer Front Left
screw1: 4, 54
screw2: Outer Front Right
screw2: 244, 54
screw3: Outer Back Right
screw3: 244, 294
screw4: Outer Back Left
screw4: 4, 294
screw5: Inner Front Left
screw5: 84, 135
screw6: Inner Front Right
screw6: 164, 135
screw7: Inner Back Right
screw7: 164, 214
screw8: Inner Back Left
screw8: 84, 214

**Reply by bassamanator** [reply]
There can be some variance in the exact positions. I'm only 100% confident about the SV06, my unit in particular. Your values look like what's in the `sv06-plus printer.cfg`, almost exactly the same, though for the `screws_tilt_adjust` section.

EDIT: actually what you have there can **only** fit into the `screws_tilt_adjust` section, not `bed_screws`. [Source](https://www.klipper3d.org/Config_Reference.html?h=bed_screw#bed_screws).

**Reply by adrianjdebono** [reply]
Thank you very much. I think I need to do a bit more research on the bed screws.

**Reply by adrianjdebono** [reply]
Apologies now I get it. I can get the dimensions for the bed screws. What i do is level the bed, and make the min z height to be a minus value, then carefully adjust the X / Y and Z offset to position the nozzle just above the screw head. I do this without the bed itself so it can be risky but worked great for the screws_tilt_adjust
Would u like me to get some dimensions tonight?

**Reply by bassamanator** [reply]
If you're up to the task, you can do the following, create a pull request and get your `contributor` badge.
- Adjust `[screws_tilt_adjust]` with your values (you're within 1mm or so, but since you're doing it with an empty plate, I trust your values more), but keep the format that I have, please see [here](https://github.com/bassamanator/Sovol-SV06-firmware/blob/9578324800bfd20df8fa8aca0e7e8203dc256d66/printer.cfg#L192).
- Create a `[bed_screws]` section that is inline with the `[screws_tilt_adjust]` (I'm concerned about screw numbers and names here). Don't worry about including other parameters (e.g., `probe_height`), just stick to screw names and their respective positions.

This is thankless work btw, I'm the only one who will thank you 😀.

**Reply by adrianjdebono** [reply]
@bassamanator  apologies for delay. I finally got around to getting the bed_screws position (this is with the nozzle directly above the bed screws.
Hopefully this helps :)

[bed_screws]
screw1: back left
screw1: 31, 274
screw2: back right
screw2: 271, 274
screw3: center back left
screw3: 111, 194
screw4: center back right
screw4: 191, 194
screw5: center front left
screw5: 111, 114
screw6: center front right
screw6: 191, 114
screw7: front left
screw7: 31, 34
screw8: front right
screw8: 271, 34
